README.md
# geojson-random

[![CircleCI](https://circleci.com/gh/node-geojson/geojson-random.svg?style=svg)](https://circleci.com/gh/node-geojson/geojson-random)

Generate random [GeoJSON](http://geojson.org/) features.

Usable in [node.js](http://nodejs.org/) and in browsers with [browserify](http://browserify.org/).

    npm install -g geojson-random
    geojson-random

    # special fast-mode for points
    geojson-random 10000 point-stream

## api

```js
var random = require('geojson-random');
```

### `random.point(count, bbox)`

Return `count` points wrapped in a FeatureCollection.

An optional `bbox` parameter should be an array of numbers representing
a [bbox](http://geojson.org/geojson-spec.html#bounding-boxes) in WSEN order,
and if given, the point will reside within its bounds.

### `random.position(bbox?)`

Return a single GeoJSON [Position](http://geojson.org/geojson-spec.html#positions)
as a 2-element array of numbers in longitude, latitude order.

An optional `bbox` parameter should be an array of numbers representing
a [bbox](http://geojson.org/geojson-spec.html#bounding-boxes) in WSEN order,
and if given, the position will reside within its bounds.

### `random.polygon(count, num_vertices, max_radial_length, bbox)`

Return `count` polygons wrapped in a FeatureCollection.

* `num_vertices` is default `10` and is how many coordinates each Polygon
  will contain.
* `max_radial_length` is the maximum number of decimal degrees latitude
  or longitude that a vertex can reach out of the center of the Polygon.
  Default is `10`.
* `bbox` (Optional) Bounding box in [minX, minY, maxX, maxY] order.

### `random.lineString(count, num_vertices, max_length, max_rotation, bbox)`

Return `count` line strings wrapped in a FeatureCollection.

* `num_vertices` is default `10` and is how many coordinates each LineString
  will contain.
* `max_length` is the maximum number of decimal degrees that a vertex can be
  from its predecessor
  Default is `0.0001`.
* `max_rotation` is the maximum number of radians that a line segment can turn
  from the previous segment.
  Default is `Math.PI / 8`.
* `bbox` (Optional) Bounding box in [minX, minY, maxX, maxY] order. This
  parameter is only applied to the starting point of the line.


CHANGELOG.md
# Change Log

All notable changes to this project will be documented in this file. See [standard-version](https://github.com/conventional-changelog/standard-version) for commit guidelines.

<a name="0.5.0"></a>
# [0.5.0](https://github.com/node-geojson/geojson-random/compare/v0.4.0...v0.5.0) (2019-01-29)


### Features

* Added bbox as command line argument. ([#19](https://github.com/node-geojson/geojson-random/issues/19)) ([d65f6bc](https://github.com/node-geojson/geojson-random/commit/d65f6bc))



<a name="0.4.0"></a>
# [0.4.0](https://github.com/tmcw/geojson-random/compare/v0.3.0...v0.4.0) (2017-04-25)


### Features

* lineString, generates linestrings ([cc7f1e7](https://github.com/tmcw/geojson-random/commit/cc7f1e7)), closes [#13](https://github.com/tmcw/geojson-random/issues/13)



# 0.3.0

* Add pointStream method

# 0.2.2

* Fix `coordInBBBOX` function to fix the output of points within small bounding
  boxes.

# 1.0.0

* Total change to API: types are now separated into their own individual
  functions. Call `.polygon` and `.point` instead of specifying a type
  parameter.
* Add `geojsonRandom.position(bbox?)`
* Add optional `bbox` parameter to points.


test.js
var test = require('tap').test,
    geojsonRandom = require('./');

test('random.position()', function(t) {
    var nobbox = geojsonRandom.position();
    t.equal(nobbox.length, 2);
    t.ok(nobbox[0] > -180 && nobbox[0] < 180, 'lon');
    t.ok(nobbox[1] > -90 && nobbox[1] < 90, 'lat');
    t.end();
});

test('random.position(bbox)', function(t) {
    var withBbox = geojsonRandom.position([0, 0, 10, 10]);
    t.equal(withBbox.length, 2);
    t.ok(withBbox[0] >= 0 && withBbox[0] <= 10, 'lon');
    t.ok(withBbox[1] >= 0 && withBbox[1] <= 10, 'lat');
    t.end();
});

test('random.point()', function(t) {
    var randomPoints = geojsonRandom.point(100);
    t.equal(randomPoints.features.length, 100, '100 points');
    t.equal(randomPoints.features[0].geometry.type, 'Point', 'features are points');
    t.end();
});

test('random.point(bbox)', function(t) {
    var randomPoints = geojsonRandom.point(1, [50, 50, 60, 60]);
    t.equal(randomPoints.features.length, 1, '1 points');
    var withBbox = randomPoints.features[0].geometry.coordinates;
    t.ok(withBbox[0] >= 50 && withBbox[0] <= 60, 'lon');
    t.ok(withBbox[1] >= 50 && withBbox[1] <= 60, 'lat');
    t.end();
});

test('random.point(bbox zero width)', function(t) {
    var randomPoints = geojsonRandom.point(1, [50, 50, 50, 60]);
    t.equal(randomPoints.features.length, 1, '1 points');
    var withBbox = randomPoints.features[0].geometry.coordinates;
    t.equal(withBbox[0], 50, 'lon');
    t.end();
});

test('random.polygon', function(t) {
    var randomPolygons = geojsonRandom.polygon(100);
    t.equal(randomPolygons.features.length, 100, '100 polygons');
    t.equal(randomPolygons.features[0].geometry.type, 'Polygon', 'features are polygons');
    t.equal(randomPolygons.features[0].geometry.coordinates[0].length, 11, 'and have 11 positions in their outer rings');

    var randomPolygonsHi = geojsonRandom.polygon(100, 20);
    t.equal(randomPolygonsHi.features[0].geometry.coordinates[0].length, 21, 'and have 21 positions in their outer rings');
    t.end();
});

test('random.polygon with bbox', function(t) {
    var randomPolygonRing = geojsonRandom.polygon(1, 5, 5, [50, 50, 60, 60]).features[0].geometry.coordinates[0];

    randomPolygonRing.forEach(function(withBbox) {
        t.ok(withBbox[0] >= 40 && withBbox[0] <= 70, 'lon');
        t.ok(withBbox[1] >= 40 && withBbox[1] <= 70, 'lat');
    });

    t.end();
});



index.js
const from = require("from2");

module.exports = function() {
  throw new Error("call .point(), .lineString(), or .polygon() instead");
};

function position(bbox) {
  if (bbox) return coordInBBBOX(bbox);
  else return [lon(), lat()];
}

module.exports.position = position;

module.exports.point = function(count, bbox) {
  const features = [];
  for (let i = 0; i < count; i++) {
    features.push(feature(bbox ? point(position(bbox)) : point()));
  }
  return collection(features);
};

module.exports.pointStream = function(count, bbox) {
  return from.obj(function(size, next) {
    if (--count) {
      next(null, feature(bbox ? point(position(bbox)) : point()));
    } else {
      next(null, null);
    }
  });
};

module.exports.polygon = function(
  count,
  num_vertices,
  max_radial_length,
  bbox
) {
  if (typeof num_vertices !== "number") num_vertices = 10;
  if (typeof max_radial_length !== "number") max_radial_length = 10;
  const features = [];
  for (let i = 0; i < count; i++) {
    let vertices = [];
    const circle_offsets = Array.apply(null, new Array(num_vertices + 1)).map(
      Math.random
    );

    circle_offsets.forEach(function sumOffsets(cur, index, arr) {
      arr[index] = index > 0 ? cur + arr[index - 1] : cur;
    });
    circle_offsets.forEach(function scaleOffsets(cur) {
      cur = (cur * 2 * Math.PI) / circle_offsets[circle_offsets.length - 1];
      const radial_scaler = Math.random();
      vertices.push([
        radial_scaler * max_radial_length * Math.sin(cur),
        radial_scaler * max_radial_length * Math.cos(cur)
      ]);
    });
    vertices[vertices.length - 1] = vertices[0]; // close the ring

    // center the polygon around something
    vertices = vertices.map(vertexToCoordinate(position(bbox)));
    features.push(feature(polygon([vertices])));
  }

  return collection(features);
};

module.exports.lineString = function(
  count,
  num_vertices,
  max_length,
  max_rotation,
  bbox
) {
  if (typeof num_vertices !== "number" || num_vertices < 2) num_vertices = 10;
  if (typeof max_length !== "number") max_length = 0.0001;
  if (typeof max_rotation !== "number") max_rotation = Math.PI / 8;

  const features = [];
  for (let i = 0; i < count; i++) {
    const startingPoint = position(bbox);
    const vertices = [startingPoint];
    for (let j = 0; j < num_vertices - 1; j++) {
      const priorAngle =
        j === 0
          ? Math.random() * 2 * Math.PI
          : Math.tan(
              (vertices[j][1] - vertices[j - 1][1]) /
                (vertices[j][0] - vertices[j - 1][0])
            );
      const angle = priorAngle + (Math.random() - 0.5) * max_rotation * 2;
      const distance = Math.random() * max_length;
      vertices.push([
        vertices[j][0] + distance * Math.cos(angle),
        vertices[j][1] + distance * Math.sin(angle)
      ]);
    }
    features.push(feature(lineString(vertices)));
  }

  return collection(features);
};

function vertexToCoordinate(hub) {
  return function(cur) {
    return [cur[0] + hub[0], cur[1] + hub[1]];
  };
}

function rnd() {
  return Math.random() - 0.5;
}
function lon() {
  return rnd() * 360;
}
function lat() {
  return rnd() * 180;
}

function point(coordinates) {
  return {
    type: "Point",
    coordinates: coordinates || [lon(), lat()]
  };
}

function coordInBBBOX(bbox) {
  return [
    Math.random() * (bbox[2] - bbox[0]) + bbox[0],
    Math.random() * (bbox[3] - bbox[1]) + bbox[1]
  ];
}

function polygon(coordinates) {
  return {
    type: "Polygon",
    coordinates: coordinates
  };
}

function feature(geom) {
  return {
    type: "Feature",
    geometry: geom,
    properties: {}
  };
}

function collection(f) {
  return {
    type: "FeatureCollection",
    features: f
  };
}

function lineString(coordinates) {
  return {
    type: "LineString",
    coordinates: coordinates
  };
}
